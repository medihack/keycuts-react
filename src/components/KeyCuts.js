import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Keys } from 'keycuts'

class KeyCuts extends Component {
  componentDidMount() {
    const options = this.props.options
    this.keys = new Keys(this._el, options)

    this.updateBindings()
    this.updateWatchers()

    this.pauseResume()
  }

  componentWillUnmount() {
    this.keys.stop()
    this.keys.reset()
  }

  componentDidUpdate(prevProps) {
    if (prevProps.bind !== this.props.bind) this.updateBindings()
    if (prevProps.watch !== this.props.watch) this.updateWatchers()
    if (prevProps.paused !== this.props.paused) this.pauseResume()
  }

  pauseResume = () => {
    if (this.props.paused) this.keys.stop()
    else this.keys.resume()
  }

  updateBindings = () => {
    this.keys.unbindAll()

    const bindings = this.props.bind

    if (!bindings) return

    if (typeof bindings !== 'object')
      throw new Error('Invalid bind parameter value: ' + bindings)

    for (let binding of Object.entries(bindings)) {
      const [shortcut, callback] = binding
      this.keys.bind(shortcut, callback)
    }
  }

  updateWatchers = () => {
    this.keys.unwatchAll()

    const callbacks = this.props.watch

    if (!callbacks) return

    if (!Array.isArray(callbacks))
      throw new Error('Invalid watch parameter value: ' + callbacks)

    for (let callback of callbacks) {
      this.keys.watch(callback)
    }
  }

  render() {
    return <div ref={el => (this._el = el)}>{this.props.children}</div>
  }
}

KeyCuts.propTypes = {
  children: PropTypes.element,
  options: PropTypes.object,
  bind: PropTypes.object,
  watch: PropTypes.array,
  paused: PropTypes.bool
}

export default KeyCuts
