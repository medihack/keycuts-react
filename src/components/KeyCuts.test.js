/* eslint-env jest */

import React from 'react'
import renderer from 'react-test-renderer'

import KeyCuts from './KeyCuts'

describe('Keycuts without children', () => {
  test('should be created successfully', () => {
    const component = renderer.create(<KeyCuts />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should have an attached keys object', () => {
    const component = renderer.create(<KeyCuts />)
    const keys = component.getInstance().keys
    expect(keys).not.toBeNull()
  })

  test('should bind handler', () => {
    const component = renderer.create(
      <KeyCuts
        bind={{
          'a + b': () => {}
        }}
      />
    )
    const keys = component.getInstance().keys
    expect(keys._handlers.length).toBe(1)
  })

  test('should add watcher', () => {
    const component = renderer.create(<KeyCuts watch={[() => {}]} />)
    const keys = component.getInstance().keys
    expect(keys._watchers.length).toBe(1)
  })
})

describe('Keycuts without children', () => {
  test('should be created successfully', () => {
    const component = renderer.create(
      <KeyCuts>
        <input />
      </KeyCuts>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
