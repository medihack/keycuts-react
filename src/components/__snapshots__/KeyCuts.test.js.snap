// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keycuts without children should be created successfully 1`] = `<div />`;

exports[`Keycuts without children should be created successfully 2`] = `
<div>
  <input />
</div>
`;
